AGGREGATE.md

```diff
@@ -246,18 +241,71 @@ utilize techniques like retries to minimize data loss.
 
 
-* The `shared_info` will be a serialized JSON object. The authenticated data for
-  decryption will consist of this string (encoded as UTF-8) with a constant
-  prefix added for domain separation. The string therefore must be forwarded to
-  the aggregation service unmodified. The reporting origin can parse the string
-  to access the encoded fields.
+* The `shared_info` will be a serialized JSON object. This exact string is used
+  as authenticated data for decryption, see [below](#encrypted-payload). The
+  string therefore must be forwarded to the aggregation service unmodified. The
+  reporting origin can parse the string to access the encoded fields.
 
 
 Note: if [debugging](https://github.com/WICG/conversion-measurement-api/blob/main/EVENT.md#optional-extended-debugging-reports)
 is enabled, additional debug fields will be present in aggregatable reports,
 including the cleartext payloads, to allow downstream systems to verify that
 reports are constructed correctly.
 
+#### Encrypted payload
+The `payload` should be a [CBOR](https://cbor.io) map encrypted via
+[HPKE](https://datatracker.ietf.org/doc/draft-irtf-cfrg-hpke/) and then base64
+encoded. The map will have the following structure:
+
+```jsonc
+// CBOR
+{
+  "operation": "histogram",  // Allows for the service to support other operations in the future
+  "reporting_origin": "https://reporter.example",
+  "data": [{"bucket": <bucket, encoded as a big-endian bytestring>, "value": <value, as an integer> }, ...]
+}
+```
+Optionally, the browser may encode multiple contributions in the same payload;
+this is only possible if all other fields in the report/payload are identical
+for the contributions.
+
+This encryption should use [AEAD](https://en.wikipedia.org/wiki/Authenticated_encryption)
+to ensure that the information in `shared_info` is not tampered with, since the
+aggregation service will need that information to do proper replay protection.
+The authenticated data will consist of the `shared_info` string (encoded as
+UTF-8) with a constant prefix added for domain separation.
+
+The encryption will use public keys specified by the aggregation service. The
+browser will encrypt payloads just before the report is sent by fetching the
+public key endpoint with an un-credentialed request. The processing origin will
+respond with a set of keys which will be stored according to standard HTTP
+caching rules, i.e. using Cache-Control headers to dictate how long to store the
+keys for (e.g. following the [freshness
+lifetime](https://datatracker.ietf.org/doc/html/rfc7234#section-4.2)). The
+browser could enforce maximum/minimum lifetimes of stored keys to encourage
+faster key rotation and/or mitigate bandwidth usage. The scheme of the JSON
+encoded public keys is as follows:
+
+```jsonc
+{
+  "keys": [
+    {
+      "id": "[arbitrary string identifying the key (up to 128 characters)]",
+      "key": "[base64 encoded public key]"
+    },
+    // Optionally, more keys.
```

Comment:
Why do we need multiple keys? How will the browser choose which key to use? Will we have any limits on the # of keys?

Reply:
Added some discussion along these lines.